### ISSUE TITLE
Truecolor in Alacritty on Windows 10

### ISSUE THREAD

**Issue by zaayer** [alacritty/alacritty#1792]
**Operating System**
Windows 10

**Shell**
PowerShell Core 6

**My YAML file**
[alacritty.yml](https://pastebin.com/zBehzNS6)

**Issue**
If I run nvim from within alacritty on Windows 10 and have termguicolors on in my init.vim, the entire screen gets covered in blue and no text can be read. It works without termguicolors, but that causes other issues with color schemes.

I confirmed that opening up the PowerShell Core console directly allows nvim to use termguicolors and everything displays correctly.

Below are screenshots of my terminal, and what it looks like when I open nvim.

[Terminal - nvim not running](https://i.imgur.com/yVsKam2.png)

[Terminal - nvim running](https://i.imgur.com/kcn4aAK.png)

Below is a screenshot of opening nvim with the powershell console (how it should look).

[nvim running in windows powershell console](https://i.imgur.com/AQW4zoA.png)

**Comment by chrisduerr** [comment]
This is likely a duplicate of https://github.com/jwilm/alacritty/issues/1662, or is truecolor different @zacps?

**Comment by zaayer** [comment]
> This is likely a duplicate of #1662, or is truecolor different @zacps?

You're right - I was scanning the names and thought it was unrelated. It sounds like the issue is coming from winpty-agent.exe, so stuck with this issue until that has been updated?

**Comment by chrisduerr** [comment]
Probably going to be a wontfix if it is really a winpty issue.

Waiting for @zacps to mark/close it though, just to make sure this is actually a winpty thing.

**Comment by zacps** [comment]
I mean on one hand truecolor is not possible to fix through winpty. On the other hand it shouldn't have failed that badly. I tested it and while colours didn't work I didn't get a whole blue screen.

![image](https://user-images.githubusercontent.com/9827696/48530975-292e3a00-e8ff-11e8-9d43-7374016eec8d.png)

Might be worth keeping this around until we can identify the cause of that?

**Comment by zaayer** [comment]
> I mean on one hand truecolor is not possible to fix through winpty. On the other hand it shouldn't have failed that badly. I tested it and while colours didn't work I didn't get a whole blue screen.
> 
> Might be worth keeping this around until we can identify the cause of that?

Thanks for the update, and yeah it does seem quite odd. I also forced the background to be the same in both vim and in Alacritty (thinking it may be because the called color isn't part of the 16 colors), but it still didn't work.

Please let me know if you need any further information from me to help.

**Comment by chrisduerr** [comment]
Note that this also happens with the ConPTY, see related issue here:
https://github.com/jwilm/alacritty/issues/2088

**Comment by zaayer** [comment]
> Note that this also happens with the ConPTY, see related issue here:
> #2088

Agreed. Also just a quick update on this, with the new experimental conhost service on an updated windows build (I believe it's still in the insiders only version) truecolor works fine in Alacritty on windows.

Not sure if it's worth keeping the issue open if it is expected to be fixed when MS releases their updated conhost to the general population, but just thought I'd let you know.

**Comment by chrisduerr** [comment]
@JDTX has mentioned that he was running the ConPTY and it was not working for him.

Maybe he can test again just to make sure it hasn't been fixed recently?

**Comment by JDTX** [comment]
@zaayer Can you tell me what build in specific you're using? The latest stable release is version 1809, which I have.

I see some relevant changelogs for the insider version 18317 here:
```
• Fixed ConPTY, enabling underline VT sequences to now pass correctly, allowing ConPTY-enabled apps (e.g. VSCode’s integrated terminal) to correctly receive and display underlined text

• Made ConPTY flush its output buffer before terminating ensuring apps receive all input and display correct output

• Significantly improved performance of ConPTY – perf now very close to “raw pipe”
```
src: https://blogs.windows.com/windowsexperience/2019/01/16/announcing-windows-10-insider-preview-build-18317/

**Comment by chrisduerr** [comment]
Looks like this is only an issue on WinPTY again then. Thanks for the update.

**Comment by zaayer** [comment]
> @zaayer Can you tell me what build in specific you're using? The latest stable release is version 1809, which I have.
> 
> I see some relevant changelogs for the insider version 18317 here:
> 
> ```
> • Fixed ConPTY, enabling underline VT sequences to now pass correctly, allowing ConPTY-enabled apps (e.g. VSCode’s integrated terminal) to correctly receive and display underlined text
> 
> • Made ConPTY flush its output buffer before terminating ensuring apps receive all input and display correct output
> 
> • Significantly improved performance of ConPTY – perf now very close to “raw pipe”
> 
> •  Fixed GitHub Issue 313, enabling colors above index #15 to be set via VT OSC 4
> ```
> src: https://blogs.windows.com/windowsexperience/2019/01/16/announcing-windows-10-insider-preview-build-18317/

I have uninstalled Windows now, but I was using the insider-preview version so ahead of the stable and didn't have those issues. I assume it should be fixed then with the latest version, although doesn't help with v1809 it sounds like.

**Comment by tycho** [comment]
I'm currently running 18990.1 (insider "skip ahead" branch). I've tried setting `enable_experimental_conpty_backend` to `true` or `false` and in both cases the rendering looks broken.

Alacritty 0.3.3's appearance regardless of WinPTY vs ConPTY backend:
![2019-09-25-2120-alacritty_RFQmlZnQ3h](https://user-images.githubusercontent.com/29616/65657949-92301500-dfda-11e9-80f4-0f55663a7231.png)

mintty, version distributed with Cygwin:
![2019-09-25-2121-mintty_egbyOijf3t](https://user-images.githubusercontent.com/29616/65657953-96f4c900-dfda-11e9-8898-d0ce15befd49.png)

My color test is this (modified version of the commonly used 256colors2.pl, with the true color ramp added in):
https://gist.github.com/tycho/73401cd6aed7e733fc2b97a8de1ca03e

Trace log with `enable_experimental_conpty_backend: true`:
[Alacritty-11888.log](https://github.com/jwilm/alacritty/files/3655880/Alacritty-11888.log)

Trace log with `enable_experimental_conpty_backend: false`:
[Alacritty-13452.log](https://github.com/jwilm/alacritty/files/3655882/Alacritty-13452.log)

**Comment by JDTX** [comment]
@tycho I discovered this was actually a problem with Cygwin and not a problem with Alacritty, just forgot to update this until you just replied.

The behavior can be described here:
https://conemu.github.io/en/CygwinAnsi.html

A summary: 
>The problem is that when cygwin’s applications detect real win32 console, cygwin’s core process all ANSI sequences internally instead of bypassing them to terminal.

So Cygwin processes the sequences instead of passing them on to Alacritty.

If colors truly bother you like they did me, use WSL/WSL2. Colors work beautifully with Alacritty & WSL.  Also, if you have OpenSSH installed on your Windows machine and you SSH to a Linux box, colors work fine there, because `ssh.exe` actually passes on the ANSI sequences to your terminal to process.

tldr: It's a Cygwin problem and I don't think there's anything Alacritty can do about it.

**Comment by tycho** [comment]
Hm. I wonder if the [cygwin/msys terminal connector](https://conemu.github.io/en/CygwinMsysConnector.html) can be made to work with Alacritty. Or maybe Cygwin can be fooled into thinking it's not a win32 console somehow, and spew out actual ANSI sequences? I admit I don't understand how all this stuff actually works under the hood, so I'm not sure what's feasible here.

**Comment by chrisduerr** [comment]
The solution is to switch to ConPTY.

**Comment by ckuai** [comment]
> EDIT2: Also the reason I don't use WSL/WSL2 is because I have my SSH private key on a smart card (YubiKey) and need to use the Windows build of gpg-agent in order to expose it to `ssh`. gpg-agent is PuTTY/Pageant compatible, but bridging it to WSL is enormous effort compared to Cygwin and MinGW's `ssh-pageant.exe`, which solves that problem trivially.

I have other reasons to stay with Cygwin, but if the only reason prevent u to switch to WSL is because of Cygwin has ssh-pageant as a bridge to putty's Pageant. then take a look in https://github.com/vuori/weasel-pageant, works pretty good for me.

**Comment by cup** [comment]
https://github.com/alacritty/alacritty/issues/2107

https://github.com/alacritty/alacritty/issues/1792
https://github.com/alacritty/alacritty/issues/1662
https://github.com/alacritty/alacritty/issues/2088
https://github.com/alacritty/alacritty/issues/2578

**Comment by chrisduerr** [comment]
WinPTY support has been removed.
